tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,271 @@
+        return *access_data<T>(batch, width, height, depth);
+    }
+
+    template<typename T>
+    const T& at(const cnn_size_t batch,
+                const cnn_size_t width,
+                const cnn_size_t height,
+                const cnn_size_t depth) const {
+        return *access_data<T>(batch, width, height, depth);
+    }
+
+    // Returns the pointer to a specified index in the tensor
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T* ptr(const cnn_size_t batch,
+           const cnn_size_t width,
+           const cnn_size_t height,
+           const cnn_size_t depth) {
+        return access_data<T>(batch, width, height, depth);
+    }
+    
+    template<typename T>
+    const T* ptr(const cnn_size_t batch,
+                 const cnn_size_t width,
+                 const cnn_size_t height,
+                 const cnn_size_t depth) const {
+        return access_data<T>(batch, width, height, depth);
+    }
+
+    // zero-overhead version (same performance to raw pointer access.
+    // have an assertion for out-of-range error)
+    template<typename T>
+    T& operator[] (cnn_size_t index) {
+        return *access_data<T>(index);
+    }
+
+    template<typename T>
+    const T& operator[] (cnn_size_t index) const {
+        return *access_data<T>(index);
+    }
+
+    // this is only a proof of concept to copy data
+    // from one device to another.
+    void toDevice(const Device& device) {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+        CLCudaAPI::Context ctx = device.context();
+        CLCudaAPI::Queue queue = device.queue();
+
+        device_data_ = buffer_ptr(new buffer(
+            ctx, queue, host_data_->begin(), host_data_->end()));
+#endif
+    }
+
+    // this is only a proof of concept to copy data
+    // from one device to another.
+    void fromDevice(const Device& device) {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+	CLCudaAPI::Queue queue = device.queue();
+
+        device_data_->Read(
+	    queue, host_data_->size(), &host_data_->at(0));
+#endif
+    }
+
+    /*template<typename T, typename U = float_t>
+    T host_data() const {
+        return static_cast<T>(*access_data(0));
+    }
+
+    template<typename T, typename U = float_t>
+    T mutable_host_data() {
+        return static_cast<T>(*access_data(0));
+    }
+
+    template<typename T, typename U = float_t>
+    T device_data() const {
+	fromDevice(device_);
+        return static_cast<T>(*access_data(0));
+    }
+
+    template<typename T, typename U = float_t>
+    T mutable_device_data() {
+	fromDevice(device_);
+        return static_cast<T>(*access_data(0));
+    }*/
+
+ private:
+    // Initializes the data buffer with zeroes
+    void init_data(const cnn_size_t batch,  const cnn_size_t width,
+                   const cnn_size_t height, const cnn_size_t depth) {
+        host_data_ = data_ptr(new std::vector<U>(
+            batch * width * height * depth, float_t(0.0)));
```

Comment:
what do you mean here? probably this method could be public and  implemented a resize routine